keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,858 @@
+
+### API Changes
+
+#### CSINode
+
+The `CSINode.Spec.Drivers[*].Allocatable.Count` field will be made mutable. No changes to the object structs are
+needed, only the validation logic needs to be revised. For reference, these are the API fields this KEP proposes to
+make mutable.
+
+```golang
+// CSINodeDriver holds information about the specification of one CSI driver installed
+type CSINodeDriver struct {
+    ...
+    // allocatable represents the volume resources of a node that are available for sc
+    // +optional
+    Allocatable *VolumeNodeResources
+}
+```
+
+```golang
+// VolumeNodeResources is a set of resource limits for scheduling of volumes.
+type VolumeNodeResources struct {
+    // Maximum number of unique volumes managed by the CSI driver that can be used on
+    // A volume that is both attached and mounted on a node is considered to be used o
+    // The same rule applies for a unique volume that is shared among multiple pods on
+    // If this field is not specified, then the supported number of volumes on this no
+    // +optional
+    Count *int32
+}
+```
+
+#### CSIDriver
+
+A new field, `AllocatableUpdateInterval`, will be added to the `CSIDriverSpec` struct. This field allows a CSI driver to
+specify the interval at which the Kubelet should periodically query a driver's `NodeGetInfo` RPC endpoint to update the
+`CSINode` object. If this field is not set, updates will only occur in response to volume attachment failures as a result of no
+capacity.
+
+```golang
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ...
+    // AllocatableUpdateInterval specifies the interval between periodic updates of
+    // the CSINode allocatable capacity for this driver. If not set, periodic updates
+    // are disabled, and updates occur only upon detecting capacity-related failures.
+    // +optional
+    AllocatableUpdateInterval *metav1.Duration
+}
+```
+
+#### Validation Changes
+
+The `ValidateCSINodeUpdate` function in the API validation code path will be modified to allow updates to the `Allocatable.Count`
+field when the feature gate is enabled:
+
+```golang
+func ValidateCSINodeUpdate(new, old *storage.CSINode) field.ErrorList {
+    allErrs := ValidateCSINode(new)
+    
+    if utilfeature.DefaultFeatureGate.Enabled(features.MutableCSINode) {
+        for _, oldDriver := range old.Spec.Drivers {
+           for _, newDriver := range new.Spec.Drivers {
+                // Allow Allocatable.Count to be modified
+                // Ensure all other fields are unchanged
+            }
+        }
+    } else {
+        // Existing validation logic for when feature gate is disabled
+    }
+    return allErrs
+}
+```
+
+This updated logic allows the `Allocatable.Count` field to be modified when the feature gate is enabled, while ensuring all
+other fields remain immutable. When the feature gate is disabled, it falls back to the existing validation logic for backward
+compatibility.
+
```

Comment:
Another validation is the entities that can update the field, only the kubelet should update it and not end users, so that might mean a change in the kube-apiserver as well.

Reply:
Hmm, I've been thinking more about this... if I understand the ask correctly, it is to ensure that **only** kubelet can update `CSINode` objects, and prevent any other entity - including privileged users - from updating this object.

It seems that the node authorizer is designed to authorize API requests _made by kubelets_, not restrict requests from other sources. In other words, no change on the node authorizer could possibly address @mauriciopoppe's request (and, no changes are needed on the node authorizer side because it already allows updating the `CSINode` object, as per @aojea's finding above : )

I'm curious to know if there is precedent before we continue down this path. I believe we don't only allow kubelet to create `CSINode` objects, wouldn't it make sense to have also enforced this restriction for object creation as well? It seems odd to allow privileged entities the ability to create but not update.

Reply:
I am not aware of any APIs where we restrict the users. The only distinction is cluster-level API vs namespace-scoped APIs.